client/src/Services/useSocketCommunication.ts

```diff
@@ -8,26 +8,35 @@ import {
 import useSocketClientStore from './socketClient';
 import { EventHandlers } from './socketClientType';
 import { useGlobalStore } from '../visualiser-debugger/Store/globalStateStore';
+import { useUserFsStateStore } from '../visualiser-debugger/Store/userFsStateStore';
+import { useFrontendStateStore } from '../visualiser-debugger/Store/frontendStateStore';
 
+interface Task {
+  execute: Promise<boolean>;
+  resolve: (result: boolean) => void;
+  id: number;
+}
+
+let initialized = false;
 export const useSocketCommunication = () => {
   const { updateNextFrame, updateTypeDeclaration, clearTypeDeclarations, clearUserAnnotation } =
     useGlobalStore();
+  const { setActive } = useFrontendStateStore();
+  const { clearFrontendState } = useFrontendStateStore();
 
   const { socketClient } = useSocketClientStore();
-  const [activeSession, setActiveSession] = useState<boolean>(false);
   const [consoleChunks, setConsoleChunks] = useState<string[]>([]);
   const { updateCurrFocusedTab } = useGlobalStore();
 
-  useEffect(() => {
+  const [, setTaskQueue] = useState<Task[]>([]);
+  const [taskId, setTaskId] = useState(0);
+
+  if (!initialized) {
```

Comment:
can use `useRef` or `useMemo` to create this functions only on first render

gpt4o:

```js
import React, { useMemo } from 'react';

const MyComponent = () => {
  const functions = useMemo(() => ({
    func1: () => { /* function logic */ },
    func2: () => { /* function logic */ },
    // Add more functions as needed
  }), []);

  const { func1, func2 } = functions;

  return (
    <div>
      {/* Use your functions here */}
      <button onClick={func1}>Run Function 1</button>
      <button onClick={func2}>Run Function 2</button>
    </div>
  );
};

export default MyComponent;
```

- useRef returns a mutable ref object whose .current property is initialized to the argument passed to useRef (in this case, null).
- The object returned by useRef will persist for the full lifetime of the component.
- By checking if functionsRef.current is null, you can ensure the functions are only created once.

Reply:
The useMemo hook in React doesn't create a single, always-the-same instance (which is what a "singleton" pattern does). Instead, useMemo is more like a "scoped" function that only runs once per component, but it doesn't guarantee that the instance (in this case, a socket) will be the same every time the component re-renders.

In practice, useSocketCommuication were used in multiple place and thus re-renders in multiple palces.